### ISSUE TITLE
could softmax_cross_entropy_with_logits's  label  has many 1.

### ISSUE THREAD

**Issue by bringtree** [tensorflow/tensorflow#15079]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Mac 10.13.1
- **TensorFlow installed from (source or binary)**:pip
- **TensorFlow version (use command below)**:v1.3.0-rc1-2456-g7abd587 1.4.0-dev20170922
- **Python version**: 3.6.1
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.
```
import numpy as np
import tensorflow as tf

a = np.array([0.0, 1.0, 0.0, 0.0, 0.0, 0.0])
# b = np.array([0.0, 1.0, 0.0, 0.0, 0.0, 0.0])  
b = np.array([0.0, 1.0, 1.0, 0.0, 1.0, 0.0])
sess = tf.Session()
print(str(sess.run(tf.nn.softmax_cross_entropy_with_logits(labels=b, logits=a))))
```
I think it maybe should throw a warn. Because the softmax_cross_entropy_with_logits's labels should not accept the label which has many 1.

If we should fix it. could you let me try to fix it ?I want to try, thanks.
### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by facaiy** [comment]
I think in most cases only one class is expected for classification problem, however multiple labels are valid for label problem. Hence it's not bug for me.

**Comment by bringtree** [comment]
@facaiy  but it should't be np.array([0.0, 1.0, 1.0, 0.0, 1.0, 0.0]) . When we need express multiple labels. the b is np.array([0.0, 0.3333333333333333, 0.3333333333333333, 0.0, 0.3333333333333333, 0.0])?

**Comment by facaiy** [comment]
I mean that multiple ones can exits in `label`, not `logit`. However, I agree that `softmax` is designed for classification problem. As for multiple labels, people prefer to using `sigmoid`, https://stats.stackexchange.com/questions/207794/what-loss-function-for-multi-class-multi-label-classification-tasks-in-neural-n

Hence your proposal sounds reasonable for me: precheck label and throw a warning that multiple ones exist.

**Comment by facaiy** [comment]
Oh, I checked the API of [` tf.nn.softmax_cross_entropy_with_logits `](https://www.tensorflow.org/versions/master/api_docs/python/tf/nn/softmax_cross_entropy_with_logits). 
> NOTE: While the classes are mutually exclusive, their probabilities need not be. All that is required is that each row of labels is a valid probability distribution. If they are not, the computation of the gradient will be incorrect.
>
> If using exclusive labels (wherein one and only one class is true at a time), see sparse_softmax_cross_entropy_with_logits.

In all,
+ ` tf.nn.softmax_cross_entropy_with_logits`： `label` must be valid probability distribution;
+ ` tf.nn.sparse_softmax_cross_entropy_with_logits`: `label` must be class id.

Hence, if necessary, we should check that label must be nonnegative and their sum equals one.

```python
import numpy as np
import tensorflow as tf

logits = np.array([[0.0, 1.0, 0.0, 0.0, 0.0, 0.0]])
labels = np.array([[0.0, 1.0, 1.0, 0.0, 1.0, 0.0]])    # invalid
loss = tf.nn.softmax_cross_entropy_with_logits(labels=labels, logits=logits)
with tf.Session() as sess:
    print(sess.run(loss))
    # [ 5.13077533]
```

**Comment by reedwm** [comment]
This issue is mentioned in a TODO [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_ops.py#L1765).

/CC @ebrevdo what do you think of this proposal? I don't think we can cause an error due to backwards compatibility, and we typically do not print warnings due to incorrect arguments. Also there would be a performance cost.

**Comment by ebrevdo** [comment]
The documentation clearly states that the input should be a probability distribution.  It seems fine to add the check for the CPU version, without worrying about backwards compatibility guarantees (though we'd want to run internal tests before accepting the change).  However, on the GPU this check is sufficiently expensive that we can't implement it without a lot of trickery and, probably, some slowdown.  So my suggestion is:

Add the check for the CPU impl, run all unit tests; then we run all internal unit tests and see if there are basic models that rely on broken behavior.

**Comment by ebrevdo** [comment]
(the check should raise `errors::InvalidArgument` if the input is not sufficiently close to a probability distribution.  probably you can't check exact, but you can check the sum is w/in e.g. 10* std::numeric_limits<T>::epsilon() of 1.

**Comment by facaiy** [comment]
Hi, @ebrevdo . I checked the implementation of `SoftmaxCrossEntropyWithLogits`: kernel functions in `xent_op` seems totally shared by both CPU and GPU. So I'm afraid that only checking in CPU impl is impossible.

Is it good to check in python side, for example, using  `tf.reduce_sum`?  Take account of performance, perhaps we'd better make it optional.

**Comment by ebrevdo** [comment]
The cpu kernel can run `if (std::is_same<Device, CPUDevice>::value) {
<perform check> }`.

**Comment by facaiy** [comment]
Thanks, @ebrevdo .
Hi, @bringtree, contributions are welcome.

**Comment by bringtree** [comment]
Could I add a arg to judge to open label's check?（Default:close the check） I think maybe some users don't want to check the sum. @ebrevdo

**Comment by bringtree** [comment]
and I find it maybe will have a problem in tf.nn.sampled_softmax_loss.  The tf.nn.sampled_softmax_loss 's labels sum maybe will not be 1(I guess). @facaiy

**Comment by bringtree** [comment]
I guess maybe it will have a problem when we use the tf.nn.sampled_softmax_loss. 
I can't be sure that  tf.nn.sampled_softmax_loss will choose the labels can be add up to 1 in future.
Though tf.nn.sampled_softmax_loss often is used to the word2vec, and In word2vec the tf.nn.sampled_softmax_loss's label sum will be 1.

**Comment by bringtree** [comment]
Sorry, I don't know how to correct it. Could you help me to correct it?  I want to know how you modify the codes?@facaiy 
`math_ops.equal(math_ops.reduce_sum.reducesum(labels),constant_op.constant(1.0,dtype=precise_logits.dtype))`

**Comment by facaiy** [comment]
@bringtree You can refer to my PR #15245 and `BinaryGradCommon` method in `cc/gradients/math_grad.cc`. Moreover, [C++ API | TensorFlow](https://www.tensorflow.org/api_guides/cc/guide) might be useful for you.

**Comment by wt-huang** [comment]
Closing as this is resolved
